[docs] add hello-dagster

slopp opened the pull request:
### Summary & Motivation

A common point of feedback is that our current docs site is great for experienced dagster users but can be intimidating for first time doc visitors. 

The goal of this addition is:

1. We often hear "dagster is complicated because it has so many concepts". Viscerally I think part of this sentiment could be explained by our doc's left nav which defaults to expanding the Concepts section. This PR instead defaults to expanding the Getting Started section. 

2. The current home landing page offers many links out to various sources of content. This PR re-arranges those links slightly to highlight our community and the most common starting point for new users, as opposed to popular articles which can be overwhelming for a first time visitor.

3. The bulk of this PR is the addition of a `hello-dagster` example that prefaces our Getting Started section. The idea of this example is to provide a single file "Copy and Paste" introduction to dagster that gets a user from 0 to dagit as quickly as possible. Just enough explanation is provided to pass on excitement about Dagster's core differentiator: assets. This starter page is really a *sales pitch for the framework*, not an explanation or a tutorial. I expect users will skim this page, internalize that dagster is as easy as 30 lines of code, and either run that code,  move into next steps, or bookmark dagster for later exploration.

### How I Tested These Changes
- [ ] code runs
- [ ] links work
- [ ]  everything reads nicely

sryza commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,99 @@
+        Get up to 500 top stories from the HackerNews topstories endpoint.
+        API Docs: https://github.com/HackerNews/API#new-top-and-best-stories
+        """
+        newstories_url = "https://hacker-news.firebaseio.com/v0/topstories.json"
```
Should `newstories_url` be `topstories_url`? Or inlined with the function call? And `top_500_newstories` be `top_500_stories`? I found myself trying to figure out what a "newstory" was.

sryza commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,99 @@
+    import requests
+    import pandas as pd
+
+    # dagster helps you think about datasets, models, 
```
Can we take this comment out?

sryza commented on docs/content/getting-started/hello-dagster.mdx:
````diff
@@ -0,0 +1,99 @@
+1. Start by creating a file called `hello-dagster.py` and copying the following:
+
+    ```python file=/getting-started/hello-dagster/hello-dagster.py
+    # copy to a file called hello-dagster.py
````
Can we take this comment out?

sryza commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,99 @@
+        df = pd.DataFrame(results)
+
+        # dagster tracks rich metadata about our assets
+        context.add_output_metadata(
```
We typically recommend adding output metadata using 
```
return Output(
    value=df,
    metadata={...},
)
```

petehunt replied:
is this documented anywhere? i have seen this pattern before but i don't think it's in our api docs anywhere

petehunt replied:
also - i wonder if we should just skip including metadata as part of the example. seems like 2 extra concepts (what's `Output`? what's `metadata`?) that we may want to introduce more gradually

slopp replied:
The documentation on output metadata is definitely in need of some love. I included it in this example so that when someone runs the pipeline they can see the hackernews data somewhere. Otherwise it is tricky to validate that dagster did anything without determining how to load the asset from a pickle file in an ugly tmp directory.

sryza commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,99 @@
+        top_500_newstories = requests.get(newstories_url).json()
+        return top_500_newstories
+
+    # dependencies between assets can be expressed as function arguments
```
Can we take this comment out? I think it's infer-able from looking at the code.

petehunt replied:
I actually think it's hard to understand what's happening. I think inferring the upstream asset from the name of the arg is fairly magical and we need to include an explanation somewhere, whether it's as as comment, or commentary after the code sample.

slopp replied:
I'll add back a comment which I think strikes a nice balance between Sandy's concerns for code pars-ability while still calling out the magical pieces.

sryza commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,99 @@
+    # dependencies between assets can be expressed as function arguments
+    @asset 
+    def hackernews_topstories(context, hackernews_topstory_ids):
+        """
```
We could make this more concise by putting the docstring on one line and removing the newline after

sryza commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,99 @@
+        for item_id in hackernews_topstory_ids:
+            item = requests.get(f"https://hacker-news.firebaseio.com/v0/item/{item_id}.json").json()
+            results.append(item)
+            if len(results) >= 10:
```
I would expect `hackernews_topstories` to contain the same stories as `hackernews_topstory_ids`. Thoughts on applying this limit in `hackernews_topstory_ids`?  It could be more concise there - just `return top_500_newstories[:min(len(top_500_newstories), 10)]`

sryza commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,99 @@
+
+## Next Steps
+
+This example uses [assets](/tutorial#assets). Most new Dagster projects use assets because they let us:
```
We typically avoid using the first person in this way, because it gets confusing whether "we/us" means "the Dagster team" (like how you're using it in "we recommend starting with assets") vs. "˝data practitioners" (like how you're using it in "Most new Dagster projects use assets because they let us".

sryza reviewed (changes requested):
I like this idea and I think the implementation is pretty close.  A few thoughts:

#### Welcome page
- I think we should stick with one main CTA on the welcome page, rather than have "Hello Dagster" and "Explore Examples" both have bold buttons. I don't have a strong opinion about which it should be.
- I think the button should have a verb, to be consistent with "View" and "Explore".

#### Hello Dagster
- I think we should see if we can make the example more concise and more self-explanatory. When I first look at it, I'm not sure where to put my attention: the comments? the definitions? It feels like there's a lot going on and I'm a tiny bit overwhelmed. Feel free to push back, but I think we could basically remove all the comments without sacrificing clarity.  I'm leaving a few granular comments on this inline.

slopp commented:
Thanks for the thoughtful feedback @sryza , I made changes to reflect your comments and this is ready to go (pending re-approval)

yuhan reviewed (approved):
stamp to unblock the landing - i think it's a great improvement. 

@erinkcochran87 added you as a reviewer to get this on your radar and would like to get your feedback especially the CTA copies on the homepage.

@slopp i pushed a commit to fix the sidebar expansion logic so now the Getting Started section is actually isDefaultOpen

erinkcochran87 commented on docs/content/getting-started/hello-dagster.mdx:
````diff
@@ -0,0 +1,95 @@
+pip install dagster dagit pandas
+```
+
+We'll use `dagit`, the dagster user interface, to run `hello-dagster.py`:
````
A few things here:

```suggestion
Using Dagit, the Dagster user interface, run `hello-dagster.py`:
```

- Dagit and Dagster should be capitalized. I think here you were referring to the command to start Dagit, but the sentence reads like it's actually the name / proper noun.

   You could say something like "Next, run `dagit` in your terminal to start Dagit, Dagster's user interface" and that would work.
- The instructions should clarify that the user needs to run this command in the same location as `hello-dagster.py`. I think it's implied, but given the number of issues beginners have I think being overly specific is fine / preferred

slopp replied:
```suggestion
Next, run`dagit` in your terminal in the same directory as `hello-dagster.py`. This command starts Dagster's user interface.
```

erinkcochran87 commented on docs/content/getting-started/hello-dagster.mdx:
````diff
@@ -0,0 +1,95 @@
+    return [hackernews_top_story_ids, hackernews_top_stories]
+```
+
+Now install the Python packages we'll need to run `hello-dagster.py` in your favorite Python environment. Unsure? Follow our [installation guide](/getting-started/install).
````
```suggestion
Install the Python packages you'll need to run `hello-dagster.py` in your favorite Python environment. Unsure? Follow our [installation guide](/getting-started/install).
```

Suggesting we change this from `we` to `you`. It read a little bit odd to me.

erinkcochran87 commented on docs/content/getting-started/hello-dagster.mdx:
````diff
@@ -0,0 +1,95 @@
+dagit -f hello-dagster.py
+```
+
+Navigate your browser to `http://localhost:3000/`. Clicking `Materialize All` will run our pipeline and create our assets.
````
```suggestion
In your browser, navigate to [http://localhost:3000/](http://localhost:3000). Click **Materialize All** to run the pipeline and create your assets.
```

Suggesting that `localhost` be a link, so users can just click it if they want to. Also changed the formatting for the button text to bold to be consistent with the style changes I'm making (but aren't necessarily in the style guide yet, sorry 😬 )

erinkcochran87 commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,95 @@
+height={874}
+/>
+
+The `hackernews_top_stories` asset included `metadata`. Click on the asset and then `Show Markdown` to preview the pipeline result and view the top HackerNews stories.
```
This is a really cool part of the example that feels sorta odd in its current state. Perhaps something like this?:

```suggestion
That's it! You now have two materialized Dagster assets. 

But wait - there's more. Because the `hackernews_top_stories` asset specified `metadata`, you can view the metadata right in Dagit.

Click the asset. In the sidebar that displays, click the **Show Markdown** link in the **Materialization in Last Run** section. This opens a preview of the pipeline result, allowing you to view the top 10 HackerNews stories:
```

I think it'd be cool to show an image of the preview here as well, or a GIF that goes from clicking **Show Markdown** to the preview opening. It admittedly took me a moment to locate the link in the UI, and I'd like to reduce the amount of thinking folks have to do.

erinkcochran87 commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,95 @@
+
+The `hackernews_top_stories` asset included `metadata`. Click on the asset and then `Show Markdown` to preview the pipeline result and view the top HackerNews stories.
+
+## Next Steps
```
```suggestion
---

## Next steps
```

Headings should use sentence case, and H2s should be preceded by a horizontal rule. It helps break things up visually

erinkcochran87 commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,95 @@
+
+This example uses [assets](/tutorial#assets). Most new Dagster projects use assets because they let data engineers:
+
+- think in the same terms as stakeholders
+- answer questions about data quality and lineage
+- work with the modern data stack (dbt, Airbyte/Fivetran, Spark)
+- create declarative freshness policies instead of task-driven cron schedules
```
```suggestion
- Think in the same terms as stakeholders
- Answer questions about data quality and lineage
- Work with the modern data stack (dbt, Airbyte/Fivetran, Spark)
- Create declarative freshness policies instead of task-driven cron schedules
```

erinkcochran87 commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,95 @@
+
+## Next Steps
+
+This example uses [assets](/tutorial#assets). Most new Dagster projects use assets because they let data engineers:
```
What's the thinking for linking to the tutorial over a different page?

slopp replied:
No strong opinion on the specific next step link. I think once the new hackernews full fledged example / tutorial is done that would be a great next step that builds off of the sample code

erinkcochran87 replied:
Definitely agreed there. Given that we're hoping to release that soon, I think this is okay to leave as-is

erinkcochran87 commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,95 @@
+- work with the modern data stack (dbt, Airbyte/Fivetran, Spark)
+- create declarative freshness policies instead of task-driven cron schedules
+
+Dagster also offers [ops and jobs](/tutorial#ops-and-jobs), but we recommend starting with assets.
```
Is the expectation that users should know what these things are? This personally feels out of place, but that might just be me

slopp replied:
My goal was to clarify that most Dagster projects should start with assets, since a very common question I receive is "should I use assets or ops?"

erinkcochran87 replied:
Yep; I get that. I was thinking about this as more of "what are ops and jobs and why should I care?" but perhaps users coming in will already have some idea what these are?

erinkcochran87 commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,95 @@
+
+Dagster also offers [ops and jobs](/tutorial#ops-and-jobs), but we recommend starting with assets.
+
+Our `hello-dagster` example used a single file, but most Dagster projects are organized as Python packages. Check out our [getting started guide](/getting-started/create-new-project) to begin with a scaffolded blank project or start from an official example such as this [dagster + dbt project](/integrations/dbt/using-dbt-with-dagster).
```
```suggestion
While this example used a single file, most Dagster projects are organized as Python packages. From here, you can:

- Start with a scaffolded blank project. Check out the [new project guide](/getting-started/create-new-project) for more info.
- Start with an official example, such as the [dbt + Dagster project](/integrations/dbt/using-dbt-with-dagster). Check out [all the examples in GitHub](https://github.com/dagster-io/dagster/tree/master/examples).
```

Suggesting we break this up a bit and put the choices into a list to make them easier to parse. I also added a link to the examples repo in GH in case folks want to peruse.

erinkcochran87 commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,95 @@
+
+# Hello Dagster
+
+This guide helps you build a simple data pipeline in dagster that downloads the top 10 HackerNews stories.
```
I like that the intro tells me what I can expect to have by the end, but this doesn't draw me in. How long will this take me? Why should I do it?

slopp replied:
```suggestion
This guide helps you build a simple data pipeline in dagster that downloads the top 10 HackerNews stories. In less than 5 minutes you will be able to develop functional code and explore dagster's user interface.
```

erinkcochran87 reviewed (commented):
Overall I dig this - thanks Sean! Left some comments, mostly about style.

slopp commented on docs/content/getting-started/hello-dagster.mdx:
````diff
@@ -0,0 +1,104 @@
+pip install dagster dagit pandas
+```
+
+Next, run`dagit` in your terminal in the same directory as `hello-dagster.py`. This command starts Dagster's user interface.
````
```suggestion
Next, run `dagit` in your terminal in the same directory as `hello-dagster.py`. This command starts Dagster's user interface.
```

slopp commented:
@erinkcochran87  thank you for the thoughtful review, we're styling! (ok... bad pun... but its late). I think I addressed all of your feedback except breaking this up into a list, I could use your help on the formatting

erinkcochran87 commented on docs/content/getting-started/hello-dagster.mdx:
```diff
@@ -0,0 +1,95 @@
+
+This guide helps you build a simple data pipeline in dagster that downloads the top 10 HackerNews stories.
+
+Start by creating a file called `hello-dagster.py` and copying the following:
```
Happy to do it!

erinkcochran87 commented:
@slopp Back to you! Here's what I did:

- I saw what you meant about the code blocks breaking the list formatting. Markdown is both a blessing and a curse. -_- Instead of using a list as I'd intended, I grouped everything into three smaller steps and added H2s. Now everything is visually broken up, and users can click around easily between steps.
- Added some updated screen shots. I noticed one of them looked stretched, and another (the one with the materialized assets) didn't fit the copy that preceded it. (That copy was telling users to materialize the assets, but the image following showed them as materialized. Thought that was confusing, so I replaced it with one that showed them in an unmaterialized state.) In the unmaterialized asset image, I also highlighted the **Materialize All** button to make it easy to find. Opted for our blurple color as the highlight since it showed up the best in color blindness tests.
- I added a "Make sure you install Python" sentence to the first paragraph. It might seem like overkill, but it's good to be specific. Another thing is that the guide uses `pip` for installation, without offering an alternative to those who may not use it. Perhaps call this out as something like _"We're using pip, but you can use whatever dependency manager you like."_

I think that's it. Please feel free to push back if you have thoughts on any of the above! Thanks again for doing this :)

sryza reviewed (approved):
This is awesome.

I have a couple of nitpicks that I think are worth addressing, but a separate PR after this release is fine.
- # asset dependencies can be inferred from parameter names would be more precise
- we should be consistent on whether there's an empty line after function docstring blocks - would lean towards no empty line
we should quickly explain what "materialize" means - i.e. that these assets are stored in pickled files on local disk
the metadata renders kind of ugly-ly. I think it's worth picking metadata that will look good.
